src/DataStructures/VectorImpl.hpp

```diff
@@ -552,3 +552,47 @@ std::ostream& operator<<(std::ostream& os,
     }                                                    \
   };                                                     \
   }  // namespace MakeWithValueImpls
+
+// {@
+/*!
+ * \ingroup DataStructuresGroup
+ * \ingroup TypeTraitsGroup
+ * \brief Helper struct to determine the base storage type of a VectorImpl or
+ * container of VectorImpl
+ *
+ * \details Extracts the storage type of a `VectorImpl`, a std::array of
+ * `VectorImpl`, or a reference or pointer to a `VectorImpl`. In any of these
+ * cases, the `type` member is defined as the `ElementType` of the `VectorImpl`
+ * in question. If, instead, `vector_base_type` is passed a numeric type, the
+ * `type` member is defined as that numeric type.
+ *
+ * \snippet DataStructures/Test_VectorImpl.cpp vector_base_type_example
+ */
+template <typename T,
+          bool = cpp17::is_arithmetic_v<T> or tt::is_a_v<std::complex, T>>
+struct vector_base_type {};
```

Comment:
You can just have the default case be the `true` one :) So:
```cpp
struct vector_base_type {
  using type = T;
};
```

Reply:
I intentionally added the true branch so that `vector_base_type_t<std::vector<double>>` is compiler error, not `std::vector<double>`.

Reply:
Ah okay. Do you still get the right behavior if you just forward declare (no ` {};` but `;`)?

Reply:
I would favor just a forward declaration (if it works correctly) because that's a common pattern

Reply:
yep, that works :+1:. The change of definition to declaration has also now been squashed in.